lib/WWW/Mechanize.pm

```diff
@@ -1770,21 +1813,45 @@ instead.)
 When given more than one pair, all criteria must match.
 Using C<undef> as value means that the attribute in question must not be present.
 
+By default, the first form that matches all criteria will be returned.
+
+    my $form = $mech->form_with( name => 'order_form', method => 'POST' );
+
+If you want the second, third or nth match, pass an optional arguments hash
+reference as the final parameter with a key C<n> to pick which instance you
+want.
+
+    my $form = $mech->form_with( method => 'POST', { n => 4 } );
+
+If the C<n> parameter is not passed, and there is more than one form on the page
+matching these criteria, then the first one is used, and a warning is generated.
+
 If it is found, the form is returned as an L<HTML::Form> object and set internally
 for later used with Mech's form methods such as
 C<L<< field()|/"$mech->field( $name, $value, $number )" >>> and
 C<L<< click()|/"$mech->click( $button [, $x, $y] )" >>>.
 
-Returns undef if no form is found.
-
+Returns C<undef> if no form is found.
 
 =cut
 
 sub form_with {
-    my ( $self, %spec ) = @_;
+    my ( $self, @args ) = @_;
 
     return if not $self->forms;
+
+    # Determine if we should return the nth instance
+    my $nth;
+    if ( @args % 2 && ref $args[-1] eq 'HASH' ) {
+        $nth = ( pop @args )->{n};
+    }
+
+    my %spec = @args;
+
     my @forms = $self->all_forms_with(%spec);
+    if ( $nth ) {
+      @forms = ( @forms >= $nth ) ? splice(@forms, $nth-1, 1) : ();
```

Comment:
nitpick: I think
```
($forms[$nth-1])
```
is more straightforward than

```
splice(@forms, $nth-1, 1)
```

Reply:
Good shout, thanks. I tried to retain as much as possible from the original commits, but this is much more readable.